onnx/defs/tensor/defs.cc

```diff
@@ -632,7 +673,29 @@ Given `data` tensor, pads, mode, and value.
     .TypeConstraint(
         "T",
         {"tensor(float16)", "tensor(float)", "tensor(double)"},
-        "Constrain input and output types to float tensors.");
+        "Constrain input and output types to float tensors.")
+    .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+        propagateElemTypeFromInputToOutput(ctx, 0, 0);
+        if (!hasNInputShapes(ctx, 1)) {
+          return;
+        }
+
+        std::vector<int64_t> pads;
+        if (!getRepeatedAttribute(ctx, "pads", pads)) {
+          return;
+        }
+
+        ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+
+        for (size_t i = 0; i < ctx.getInputType(0)->tensor_type().shape().dim_size(); ++i) {
+          auto* newdim = ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape()->add_dim();
+          if (ctx.getInputType(0)->tensor_type().shape().dim(i).has_dim_value()) {
+            newdim->set_dim_value(ctx.getInputType(0)->tensor_type().shape().dim(i).dim_value() + pads[2 * i] + pads[2 * i + 1]);
+          } else {
+            *newdim = ctx.getInputType(0)->tensor_type().shape().dim(i);
```

Comment:
This doesn't seem correct unless the pads value is zero. In particular, if input dim is "N" and pads is non-zero, we shouldn't set output dim to be "N"

Reply:
good catch